helpers/sql-obfuscation/lib/opentelemetry/helpers/sql_obfuscation.rb

```diff
@@ -0,0 +1,151 @@
+      OBFUSCATION_FAILURE_MESSAGE = 'Failed to obfuscate SQL'
+
+      # We use these to check whether the query contains any quote characters
+      # after obfuscation. If so, that's a good indication that the original
+      # query was malformed, and so our obfuscation can't reliably find
+      # literals. In such a case, we'll replace the entire query with a
+      # placeholder.
+      CLEANUP_REGEX = {
+        default: %r{'|"|\/\*|\*\/},
+        mysql: %r{'|"|\/\*|\*\/},
+        postgres: %r{'|\/\*|\*\/|\$(?!\?)}
+      }.freeze
+
+      # @api private
+      def self.generate_regex(dialect)
+        components = DIALECT_COMPONENTS[dialect]
+        Regexp.union(components.map { |component| COMPONENTS_REGEX_MAP[component] })
+      end
+
+      DEFAULT_COMPONENTS_REGEX = generate_regex(:default)
+      MYSQL_COMPONENTS_REGEX = generate_regex(:mysql)
+      POSTGRES_COMPONENTS_REGEX = generate_regex(:postgres)
+
+      # This is a SQL obfuscation utility intended for use in database adapter instrumentation.
+      #
+      # @param sql [String] The SQL to obfuscate.
+      # @param obfuscation_limit [optional Integer] The maximum length of an obfuscated sql statement.
+      # @param adapter [optional Symbol] the type of database adapter calling the method. `:default`, `:mysql` and `:postgres` are supported.
+      # @return [String] The SQL query string where the values are replaced with "?". When the sql statement exceeds the obufscation limit
+      #  the first matched pair from the SQL statement will be returned, with an appended truncation message. If trunaction is unsuccessful,
+      #  a string describing the error will be returned.
+      #
+      # @api public
+      def self.obfuscate_sql(sql, obfuscation_limit: 2000, adapter: :default)
+        regex = case adapter
+                when :mysql
+                  MYSQL_COMPONENTS_REGEX
+                when :postgres
+                  POSTGRES_COMPONENTS_REGEX
+                else
+                  DEFAULT_COMPONENTS_REGEX
+                end
+
+        sql = utf8_encode_sql(sql)
+        return truncate_statement(sql, regex, obfuscation_limit) if sql.size > obfuscation_limit
+
+        sql = sql.gsub(regex, PLACEHOLDER)
+        return UNMATCHED_PAIRS_FAILURE_MESSAGE if detect_unmatched_pairs(sql, adapter)
+
+        sql
+      rescue StandardError => e
+        OpenTelemetry.handle_error(message: OBFUSCATION_FAILURE_MESSAGE, exception: e)
+        "OpenTelemetry error: #{OBFUSCATION_FAILURE_MESSAGE}"
+      end
+
+      # @api private
+      def self.truncate_statement(sql, regex, limit)
+        first_match_index = sql.index(regex)
+        truncation_message = "SQL truncated (> #{limit} characters)"
+        return truncation_message unless first_match_index
+
+        truncated_sql = sql[..first_match_index - 1]
+        "#{truncated_sql}...\n#{truncation_message}"
+      end
+
+      # @api private
+      def self.utf8_encode_sql(sql)
```

Comment:
Is this used in more than one place? If it isn't consider inlining it because it will reduce the number of method dispatches as well as keep from adding an additional line to the call stack.

Reply:
Good point. `.utf8_encode_sql` and `.detect_unmatched_pairs` both are not used in more than one place. I'll make that change.